### PR TITLE
New annotations and dataType annotations

### PR DESCRIPTION
**Pull request by lordjim666**

### CODE REVIEW

File: Annotation/Property/AutoIncrement.php [hunk 1]
```diff
@@ -0,0 +1,39 @@
...
+     */
+    public static function name() {
+        return "autoIncrement";
+    }
+
+    /**
+     * @return array
+     */
+    public static function parametersDefinition() {
+        return array("numberOfParameters" => 1);
```
Comment (obophp) [inline Annotation/Property/AutoIncrement.php]: Number of parameters is "?"

**Comment by lordjim666** [comment]
Hm I would say that "?" should be deprecated now. It'll be always 1 parameter. 0 is not possible now.

File: Annotation/Property/Nullable.php [hunk 2]
```diff
@@ -0,0 +1,52 @@
...
+ * @link https://github.com/obophp/obo
+ * @license http://www.apache.org/licenses/LICENSE-2.0 Apache License, Version 2.0
+ */
+
+namespace obo\Annotation\Property;
+
+class Nullable extends \obo\Annotation\Base\Property {
+
+    /** @var bool */
+    public $nullable = false;
```
Comment (obophp) [inline Annotation/Property/Nullable.php]: This property must be in propertyInformation

File: Annotation/Property/AutoIncrement.php [hunk 3]
```diff
@@ -0,0 +1,39 @@
...
+    }
+
+    public function validate() {
+        if (!$this->propertyInformation->dataType instanceof \obo\DataType\Integer) throw new \obo\Exceptions\BadAnnotationException("Annotation 'autoIncrement' can be used only with 'integer' dataType.");
```
Comment (obophp) [inline Annotation/Property/AutoIncrement.php]: Data types smallInt and bigInt are not supported?

File: Annotation/Property/Nullable.php [hunk 4]
```diff
@@ -0,0 +1,52 @@
...
+    }
+
+    /**
+     * @return array
+     */
+    public static function parametersDefinition() {
+        return array("numberOfParameters" => 1);
+    }
+
+    public function validate() {
```
Comment (obophp) [inline Annotation/Property/Nullable.php]: Remove empty method pls

File: Annotation/Property/Nullable.php [hunk 5]
```diff
@@ -0,0 +1,52 @@
...
+     */
+    public static function name() {
+        return "nullable";
+    }
+
+    /**
+     * @return array
+     */
+    public static function parametersDefinition() {
+        return array("numberOfParameters" => 1);
```
Comment (obophp) [inline Annotation/Property/Nullable.php]: number of parameters should be "?"

File: Annotation/Property/Nullable.php [hunk 6]
```diff
@@ -0,0 +1,52 @@
...
+    public function process($values) {
+        parent::process($values);
+        if (\is_bool($values[0]) === false) throw new \obo\Exceptions\BadAnnotationException("Parameter for 'nullable' annotation must be of boolean type");
+        $this->nullable = $values[0];
+    }
+
+    /**
+     * @return void
+     */
+    public function registerEvents() {
```
Comment (obophp) [inline Annotation/Property/Nullable.php]: Remove empty method pls

File: DataType/Base/DataType.php [hunk 7]
```diff
@@ -30,6 +30,18 @@ public function __construct(\obo\Carriers\PropertyInformationCarrier $propertyIn
     public abstract function name();
 
     /**
+     * @param mixed $value
+     * @return void|\obo\Exceptions\Exception
+     */
+    public function validate($value) {
+        foreach ($this->propertyInformation->annotations as $annotation) {
```
Comment (obophp) [inline DataType/Base/DataType.php]: If use information from propertyInformation, you will not have to iterate annotations

File: DataType/Boolean.php [hunk 8]
```diff
@@ -25,7 +25,24 @@ public function name() {
      * @return void
      */
     public function validate($value) {
-        if (!\is_bool($value)) throw new \obo\Exceptions\BadDataTypeException("New value for property with name '{$this->propertyInformation->name}' must be boolean, " . \gettype($value) . " given");
+        parent::validate($value);
+        if ($value !== false
+           && $value !== "false"
+           && $value !== 0
```
Comment (obophp) [inline DataType/Boolean.php]: No

File: DataType/Boolean.php [hunk 9]
```diff
@@ -25,7 +25,24 @@ public function name() {
...
+           && $value !== "0"
```
Comment (obophp) [inline DataType/Boolean.php]: No

File: DataType/Boolean.php [hunk 10]
```diff
@@ -25,7 +25,24 @@ public function name() {
...
+           && $value !== true
+           && $value !== "true"
+           && $value !== 1
```
Comment (obophp) [inline DataType/Boolean.php]: No

File: DataType/Boolean.php [hunk 11]
```diff
@@ -25,7 +25,24 @@ public function name() {
...
+           && $value !== "1"
```
Comment (obophp) [inline DataType/Boolean.php]: No

File: DataType/Boolean.php (hunk 9 above)
Comment (lordjim666) [inline DataType/Boolean.php]: Yes

File: DataType/Boolean.php (hunk 8 above)
Comment (lordjim666) [inline DataType/Boolean.php]: Yes

File: DataType/Boolean.php (hunk 10 above)
Comment (lordjim666) [inline DataType/Boolean.php]: yes

File: DataType/Boolean.php (hunk 11 above)
Comment (lordjim666) [inline DataType/Boolean.php]: Yes

File: Annotation/Property/AutoIncrement.php (hunk 3 above)
Comment (lordjim666) [inline Annotation/Property/AutoIncrement.php]: Those two extends integer class

File: Annotation/Property/Nullable.php (hunk 2 above)
Comment (lordjim666) [inline Annotation/Property/Nullable.php]: Here i am not sure what you mean. Will be discussed in person

File: Annotation/Property/Nullable.php (hunk 5 above)
Comment (lordjim666) [inline Annotation/Property/Nullable.php]: "?" Has no point anymore... there will never be 0 parameters

**Pull request merged**